plugins/flytekit-kf-pytorch/tests/test_shared.py

```diff
@@ -0,0 +1,109 @@
+        # Test disabling shared memory volume
+        (PyTorch(num_workers=3, increase_shared_mem=False), None, False, False),
+        (Elastic(nnodes=2, increase_shared_mem=False), None, False, False),
+        # Test that explicitly passed pod template does not break adding shm volume
+        (Elastic(nnodes=2, increase_shared_mem=True), PodTemplate(), True, False),
+        # Test that pod template with container does not break adding shm volume
+        (
+            Elastic(nnodes=2),
+            PodTemplate(
+                pod_spec=V1PodSpec(containers=[V1Container(name="primary")]),
+            ),
+            True,
+            False,
+        ),
+        # Test that pod template with volume/volume mount does not break adding shm volume
+        (
+            Elastic(nnodes=2),
+            PodTemplate(
+                pod_spec=V1PodSpec(
+                    containers=[
+                        V1Container(name="primary", volume_mounts=[V1VolumeMount(name="foo", mount_path="/bar")])
+                    ],
+                    volumes=[V1Volume(name="foo")],
+                ),
+            ),
+            True,
+            False,
+        ),
+        # Test that pod template with multiple containers raises an error
+        (
+            Elastic(nnodes=2),
+            PodTemplate(
+                pod_spec=V1PodSpec(
+                    containers=[
+                        V1Container(name="primary"),
+                        V1Container(name="secondary"),
+                    ]
+                ),
+            ),
+            True,
+            True,
+        ),
+    ],
+)
+def test_task_shared_memory(
+    task_config: Union[Elastic, PyTorch], pod_template: PodTemplate, needs_shm_volume: bool, raises: bool
+):
+    """Test that the task pod template is configured with a shared memory volume if needed."""
+
+    expected_volume = V1Volume(name="shm", empty_dir=V1EmptyDirVolumeSource(medium="Memory"))
+    expected_volume_mount = V1VolumeMount(name="shm", mount_path="/dev/shm")
+
+    with pytest.raises(ValueError) if raises else nullcontext():
+
+        @task(
+            task_config=task_config,
+            pod_template=pod_template,
+        )
+        def test_task() -> None:
+            pass
+
+        if needs_shm_volume:
+            assert test_task.pod_template is not None
+            assert test_task.pod_template.pod_spec is not None
+            assert test_task.pod_template.pod_spec.volumes is not None
+            assert test_task.pod_template.pod_spec.containers is not None
+            assert test_task.pod_template.pod_spec.containers[0].volume_mounts is not None
+
+            assert any([v == expected_volume for v in test_task.pod_template.pod_spec.volumes])
+            assert any(
+                [v == expected_volume_mount for v in test_task.pod_template.pod_spec.containers[0].volume_mounts]
+            )
+
+        else:
+            no_pod_template = test_task.pod_template is None
```

Comment:
Can these all be asserts similar to the previous branch?

Reply:
In the previous branch we simply always assert `is not None` at every nesting level while here in this branch we ultimately want to assert that there are no shared memory volumes+mounts but in theory we could have other volumes+mounts. Or we don't have a container, or no pod spec, or no pod template so we can't just have `assert is None` at every nesting level 🤔 

If I'm overlooking a nicer way to do this, I'm super open though!

Reply:
In the parameterization, there are two entries with `needs_shm_volume=False`:

```python
        (PyTorch(num_workers=3, increase_shared_mem=False), None, False, False),
        (Elastic(nnodes=2, increase_shared_mem=False), None, False, False),
```

I would be okay with this complexity, if we add a new text case with `increase_shared_mem=False` and a `PodTemplate`.

Also can you add a comment to this branch to state that is checking that there are no shm volume mounted.

Reply:
https://github.com/flyteorg/flytekit/pull/2543/commits/d6941d5655452538f2ee79ac18f2027960c06f3d